src/main/java/com/modrinth/minotaur/TaskModrinthUpload.java

```diff
@@ -221,4 +188,72 @@ public void apply() {
+		pluginLoaderMap.put("xyz.jpenilla.run-paper", "paper");
+		pluginLoaderMap.put("xyz.jpenilla.run-waterfall", "waterfall");
+		pluginLoaderMap.put("xyz.jpenilla.run-velocity", "velocity");
+
+		pluginLoaderMap.forEach((plugin, loader) -> {
+			if (pluginManager.hasPlugin(plugin)) {
+				getLogger().debug("Adding loader '{}' because plugin '{}' was found.", loader, plugin);
+				add(ext.getLoaders(), loader);
+			}
+		});
+
+		if (!ext.getLoaders().get().contains("quilt") && getProject().getExtensions().findByName("loom") != null) {
+			String loomPlatform = getProject().getProviders().gradleProperty("loom.platform").getOrNull();
+			if (loomPlatform != null && loomPlatform.equalsIgnoreCase("forge")) {
+				getLogger().debug("Adding loader 'forge' because 'loom' extension was found and loom.platform=forge.");
+				add(ext.getLoaders(), "forge");
+			} else {
+				getLogger().debug("Adding loader 'fabric' because 'loom' extension was found.");
+				add(ext.getLoaders(), "fabric");
+			}
+		}
+	}
+
+	private void autoDetectGameVersions(ModrinthExtension ext, PluginManager pluginManager) {
+		if (ext.getLoaders().get().contains("forge") && pluginManager.hasPlugin("net.minecraftforge.gradle")) {
+			// ForgeGradle will store the game version here.
+			// https://github.com/MinecraftForge/ForgeGradle/blob/FG_5.0/src/userdev/java/net/minecraftforge/gradle/userdev/MinecraftUserRepo.java#L199
+			String version = (String) getProject().getExtensions().getExtraProperties().get("MC_VERSION");
+			assert version != null;
+
+			getLogger().debug("Adding fallback game version {} from ForgeGradle.", version);
+			add(ext.getGameVersions(), version);
+		}
+
+		if (ext.getLoaders().get().contains("fabric") || ext.getLoaders().get().contains("quilt")
+			|| (ext.getLoaders().get().contains("forge") && !pluginManager.hasPlugin("net.minecraftforge.gradle"))) {
+			if (getProject().getExtensions().findByName("loom") != null) {
+				// Use the same method Loom uses to get the version.
+				// https://github.com/FabricMC/fabric-loom/blob/97f594da8e132c3d33cf39fe8d7cc0e76d84aeb6/src/main/java/net/fabricmc/loom/configuration/DependencyInfo.java#LL60C26-L60C56
+				String version = getProject().getConfigurations().getByName("minecraft")
+					.getDependencies().iterator().next().getVersion();
+				assert version != null;
+
+				getLogger().debug("Adding fallback game version {} from Loom.", version);
+				add(ext.getGameVersions(), version);
+			}
+		}
+
+		if (ext.getLoaders().get().contains("paper") && getProject().getExtensions().findByName("paperweight") != null) {
+			final String mcVer = getProject().getExtensions().getByType(PaperweightUserExtension.class).getMinecraftVersion().get();
+			getLogger().debug("Adding fallback game version {} from paperweight-userdev.", mcVer);
+			add(ext.getGameVersions(), mcVer);
+		}
+	}
+
+	// needed because loaders is a list, not a set
+	private static  <T> void add(final ListProperty<T> list, final T element) {
```

Comment:
Do you know if it would break API if we were to change these to `SetProperty`s?

(also, there's an extra space between `static` and the generic)

Reply:
It would break API and ABI, yes. It wouldn't break most people's build scripts however because they will get recompiled.

Reply:
Hmm, I wonder if anyone would actually be affected, then... meh, still probably better to wait for v3